js/piwik.js

```diff
@@ -2484,6 +2490,66 @@ if (typeof Piwik !== 'object') {
                 }
             }
 
+            /*
+             * Sets up the heart beat timeout.
+             */
+            function heartBeatUp(delay) {
+                if (heartBeatTimeout) {
+                    return;
+                }
+
+                heartBeatTimeout = setTimeout(function heartBeat() {
+                    heartBeatTimeout = null;
+                    if (heartBeatPingIfActivityAlias()) {
+                        return;
+                    }
+
+                    var now = new Date(),
+                        heartBeatDelay = configHeartBeatDelayInSeconds - (now.getTime() - lastTrackerRequestTime);
+                    // sanity check
+                    heartBeatDelay = Math.min(configHeartBeatDelayInSeconds, heartBeatDelay);
+                    heartBeatUp(heartBeatDelay);
+                }, delay || configHeartBeatDelayInSeconds);
+            }
+
+            /*
+             * Removes the heart beat timeout.
+             */
+            function heartBeatDown() {
+                if (!heartBeatTimeout) {
+                    return;
+                }
+
+                clearTimeout(heartBeatTimeout);
+                heartBeatTimeout = null;
+            }
+
+            /*
+             * Setup event handlers and timeout for initial heart beat.
+             */
+            function setUpHeartBeat() {
+                if (heartBeatSetUp) {
+                    return;
+                }
+
+                heartBeatSetUp = true;
+
+                addEventListener(windowAlias, 'focus', function () {
+                    // since it's possible for a user to come back to a tab after several hours or more, we try to send
+                    // a ping if the page is active. (after the ping is sent, the heart beat timeout will be set)
+                    if (heartBeatPingIfActivityAlias()) {
+                        return;
+                    }
+
+                    heartBeatUp();
+                });
+                addEventListener(windowAlias, 'blur', function () {
+                    heartBeatDown();
+                });
+
+                heartBeatUp();
```

Comment:
I don't really understand this, `setUpHeartBeat()` sets up events and the first timeout, but the timeout is set via `heartBeatUp()`. `heartBeatUp()` is called in many places, including in heartBeatUp itself, because the timeout does not always exist even if the heart beat is enabled.

I will probably merge w/o this change, but if you want to clarify we can talk on slack.